File: backend/app/api/skills.py
"""Read-only metadata API for the active advisory skill catalog."""

from __future__ import annotations

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field, field_validator

from app.skills.models import SkillDescriptor
from app.skills.service import get_skill_service

router = APIRouter()


class SkillPreviewRequest(BaseModel):
    query: str = Field(min_length=1, max_length=10_000)
    flow: str = Field(default="deep_research", min_length=1, max_length=64)
    max_results: int | None = Field(default=None, ge=1, le=8)

    @field_validator("query", "flow")
    @classmethod
    def reject_blank_text(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("value cannot be blank")
        return normalized


def _serialize_skill(skill: SkillDescriptor, *, loaded: bool = False) -> dict:
    metadata = skill.metadata
    return {
        "name": metadata.name,
        "description": metadata.description,
        "version": metadata.version,
        "author": metadata.author,
        "license": metadata.license,
        "tags": list(metadata.tags),
        "category": metadata.category,
        "dependencies": list(metadata.dependencies),
        "source_path": skill.relative_path,
        "content_sha256": skill.content_sha256,
        "byte_size": skill.byte_size,
        "body_chars": skill.body_chars,
        "reference_count": len(skill.references),
        "reference_paths": list(skill.reference_paths),
        "availability": skill.availability.value,
        "blocked_reason": skill.blocked_reason or None,
        "loaded": loaded,
    }


@router.get("/status")
async def get_skill_status():
    service = get_skill_service()
    status, diagnostics = service.status_view()
    return {
        **status,
        "diagnostics": list(diagnostics),
    }


@router.get("")
async def list_skills():
    service = get_skill_service()
    status, skills = service.catalog_view()
    return {
        **status,
        "skills": [_serialize_skill(skill, loaded=loaded) for skill, loaded in skills],
    }


@router.post("/preview")
async def preview_skill_selection(request: SkillPreviewRequest):
    """Preview metadata routing without loading any skill or reference text."""

    service = get_skill_service()
    preview, status, loaded_flags = service.preview_view(
        request.query,
        flow=request.flow,
        limit=request.max_results,
    )
    selected = []
    for index, (item, loaded) in enumerate(
        zip(preview.selections, loaded_flags, strict=True),
        start=1,
    ):
        payload = _serialize_skill(
            item.skill,
            loaded=loaded,
        )
        payload.update(
            {
                "rank": index,
                "score": item.score,
                "matched_terms": list(item.matched_terms),
                "match_reason": (
                    f"Matched metadata terms: {', '.join(item.matched_terms)}"
                    if item.matched_terms
                    else "Matched skill metadata"
                ),
            }
        )
        selected.append(payload)
    cache = {
        "loaded_count": status["loaded_count"],
        "loaded_bytes": status["loaded_bytes"],
        "loaded_reference_count": status["loaded_reference_count"],
        "entry_count": status["cache_entry_count"],
        "total_bytes": status["cache_total_bytes"],
        "max_entries": status["cache_max_entries"],
        "max_bytes": status["cache_max_bytes"],
        "hits": status["cache_hits"],
        "misses": status["cache_misses"],
        "evictions": status["cache_evictions"],
    }
    return {
        "query": request.query,
        "flow": request.flow,
        "source_revision": preview.source_revision,
        "catalog_revision": preview.catalog_revision,
        "selected_count": len(selected),
        "available_count": status["available_count"],
        "loaded_count": status["loaded_count"],
        "selected": selected,
        "cache": cache,
    }


@router.get("/{name}")
async def get_skill(name: str):
    service = get_skill_service()
    skill, source_revision, catalog_revision, loaded = service.descriptor_view(name)
    if skill is None:
        raise HTTPException(status_code=404, detail="Skill not found")
    return {
        **_serialize_skill(skill, loaded=loaded),
        "source_url": service.status()["source_url"],
        "source_revision": source_revision,
        "catalog_revision": catalog_revision,
    }

File: frontend/src/api/skills.ts
import { getGuestId } from "@/store/guestStore";

const API_BASE = import.meta.env.VITE_API_URL ?? "";

export type SkillAvailability = "available" | "blocked";

export interface SkillMetadata {
  name: string;
  description: string;
  version: string;
  author: string;
  license: string;
  tags: string[];
  category: string;
  dependencies: string[];
  source_path: string;
  content_sha256: string;
  byte_size?: number;
  body_chars?: number;
  reference_count?: number;
  reference_paths?: string[];
  availability: SkillAvailability;
  blocked_reason: string | null;
  loaded: boolean;
}

export interface SkillDetailResponse extends SkillMetadata {
  source_url: string;
  source_revision: string | null;
  catalog_revision: string | null;
}

export interface SkillDiagnostic {
  code: string;
  severity: string;
  path: string;
  message: string;
}

export interface SkillsStatus {
  enabled: boolean;
  state: "disabled" | "empty" | "loading" | "ready" | "stale" | "error" | string;
  source_url: string;
  requested_ref: string;
  revision: string | null;
  catalog_revision: string | null;
  source_status: string | null;
  refreshed_at: number | string | null;
  skill_count: number;
  available_count: number;
  blocked_count: number;
  quarantined_count: number;
  diagnostic_count: number;
  error: string | null;
  cache_scope: "process" | string;
  loaded_count: number;
  loaded_bytes: number;
  loaded_reference_count: number;
  cache_entry_count: number;
  cache_total_bytes: number;
  cache_max_entries: number;
  cache_max_bytes: number;
  cache_hits: number;
  cache_misses: number;
  cache_evictions: number;
  diagnostics?: SkillDiagnostic[];
}

export interface SkillsListResponse extends SkillsStatus {
  skills: SkillMetadata[];
}

export interface SkillPreviewItem extends SkillMetadata {
  rank: number;
  score: number;
  matched_terms: string[];
  match_reason: string;
}

export interface SkillCacheSnapshot {
  loaded_count: number;
  loaded_bytes: number;
  loaded_reference_count: number;
  entry_count: number;
  total_bytes: number;
  max_entries: number;
  max_bytes: number;
  hits: number;
  misses: number;
  evictions: number;
}

export interface SkillPreviewResponse {
  query: string;
  flow: string;
  source_revision: string | null;
  catalog_revision: string | null;
  selected_count: number;
  available_count: number;
  loaded_count: number;
  selected: SkillPreviewItem[];
  cache: SkillCacheSnapshot;
}

export interface SkillPreviewRequest {
  query: string;
  flow: string;
  max_results?: number;
}

async function skillRequest<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${API_BASE}${path}`, {
    ...init,
    headers: {
      "Content-Type": "application/json",
      "X-Guest-Id": getGuestId(),
      ...(init?.headers ?? {}),
    },
  });

  if (!response.ok) {
    const detail = await response.text().catch(() => "");
    throw new Error(detail || `Skills API returned ${response.status}`);
  }

  return response.json() as Promise<T>;
}

/**
 * Isolated adapter for the skills metadata surface.
 *
 * Keeping these calls outside the main API client makes it straightforward to
 * evolve the loader contract without coupling it to research-run requests.
 */
export const skillsApi = {
  status(): Promise<SkillsStatus> {
    return skillRequest("/api/skills/status");
  },

  list(): Promise<SkillsListResponse> {
    return skillRequest("/api/skills");
  },

  detail(name: string): Promise<SkillDetailResponse> {
    return skillRequest(`/api/skills/${encodeURIComponent(name)}`);
  },

  preview(payload: SkillPreviewRequest): Promise<SkillPreviewResponse> {
    return skillRequest("/api/skills/preview", {
      method: "POST",
      body: JSON.stringify(payload),
    });
  },
};
